File: DataStructure/Queue/752_openLock.cpp
#include <vector>
#include <string>
#include <iostream>
#include <queue>
#include <set>
using namespace std;

class Solution {
public:
    string codeplus(string str, int i) {
        str[i] = str[i] == '9' ? '0' : str[i] + 1;
        return str;
    }

    string codeminus(string str, int i) {
        str[i] = str[i] == '0' ? '9' : str[i] - 1;
        return str;
    }

    int openLock(vector<string> &deadends, string target)
    {
        set<string> s;
        for (string it : deadends)
            s.insert(it);
        queue<string> q;
        q.push("0000");
        int step = 0;

        while (!q.empty()) {
            int sz = q.size();
            for (int i = 0; i < sz; i++) {
                string temp = q.front();
                q.pop();
                if (s.count(temp) != 0)
                    continue;
                if (temp == target)
                    return step;
                s.insert(temp);
                
                for (int j = 0; j < 4; j++) {
                    string tempplus = codeplus(temp, j);
                    if (s.count(tempplus) == 0) {
                        q.push(tempplus);
                    }
                    string tempminus = codeminus(temp, j);
                    if (s.count(tempminus) == 0) {
                        q.push(tempminus);
                    }
                }
            }
            step++;
        }
        return -1;
    }
};
// 终于通过了 
// str[] 里边是ascii值 这个得记住

File: DataStructure/Queue/752_openLock.java
package LeetCode.BFS;
import java.util.List;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.HashSet;
import java.util.Queue;
import java.lang.StringBuilder;

class Solution {
    public String codeplus(String str, int i) {
        StringBuilder temp = new StringBuilder(str);
        temp.replace(i, i, ((temp.charAt(i) + 1) % 10).toString());
    }

    public int openLock(String[] deadends, String target) {
        HashSet<String> set = new HashSet<>();
        for (String str : deadends)
            set.add(str);
        Queue<String> q = new LinkedList<>();
        q.offer("0000");
        int step = 0;

        while (!q.isEmpty()) {
            int sz = q.size();
            for (int i = 0; i < sz; i++) {
                String temp = q.poll();
                if (temp == target)
                    return step;
                
                for (int j = 0; j < 4; j++) {
                }
            }
        }
    }
}